src/Features/CSharp/Portable/UseIsNullCheck/CSharpUseIsNullCheckForEqualityOperatorDiagnosticAnalyzer.cs

```diff
@@ -54,43 +51,71 @@ private void AnalyzeSyntax(SyntaxNodeAnalysisContext context)
+            else
+            {
+                return;
+            }
+            if (notNull.IsKind(SyntaxKind.NullLiteralExpression))
             {
                 return;
             }
 
-            var severity = option.Notification.Severity;
-            context.ReportDiagnostic(
-                DiagnosticHelper.Create(
-                    Descriptor, binaryExpression.GetLocation(), severity, additionalLocations: null, s_properties));
-        }
+            var hasCast = false;
+            if (notNull is CastExpressionSyntax castExpression &&
+                semanticModel.GetTypeInfo(castExpression.Type).Type?.SpecialType == SpecialType.System_Object)
+            {
+                hasCast = true;
+                notNull = castExpression.Expression;
+            }
 
-        private static bool IsObjectCastAndNullCheck(
-            SemanticModel semanticModel, ExpressionSyntax left, ExpressionSyntax right)
-        {
-            if (left is CastExpressionSyntax castExpression &&
-                right.IsKind(SyntaxKind.NullLiteralExpression))
+            var exprType = semanticModel.GetTypeInfo(notNull).Type;
+            if (exprType.IsValueType)
             {
-                // make sure it's a cast to object, and that the thing we're casting actually has a type.
-                if (semanticModel.GetTypeInfo(castExpression.Type).Type?.SpecialType == SpecialType.System_Object)
+                // `t == null` can't happen.
+                // `(object)t == null` can be safely converted to `(object)t is null`.
+                // `t is null` won't be permited.
+                hasCast = false;
+            }
+            else if (!exprType.IsReferenceType) // Unconstrained type parameter
+            {
+                // Check 8.0 if https://github.com/dotnet/csharplang/issues/1284 is considered implemented.
+                if (hasCast)
                 {
-                    var expressionType = semanticModel.GetTypeInfo(castExpression.Expression).Type;
-                    if (expressionType != null)
-                    {
-                        if (expressionType is ITypeParameterSymbol typeParameter &&
-                            !typeParameter.HasReferenceTypeConstraint)
-                        {
-                            return false;
-                        }
-
-                        return true;
-                    }
+                    // if (<8.0)
+                    hasCast = false;
+                }
+                else
+                {
+                    // if (<8.0)
+                    return;
                 }
             }
 
-            return false;
+            var properties = ImmutableDictionary<string, string>.Empty.Add(
+                UseIsNullConstants.Kind, UseIsNullConstants.CastAndEqualityKey);
+            if (hasCast)
+            {
+                properties = properties.Add(CSharpUseIsNullCheckForEqualityOperatorCodeFixProvider.RemoveObjectCast, "");
+            }
```

Comment:
this condition is wrong.  we sholud leave the cast in if it's a non-nullable value type.  not if it had a cast originally.